contracts/rbn-staking/FeeCustody.sol

```diff
@@ -0,0 +1,425 @@
+
+      uint256 balance = IERC20(assets[i]).balanceOf(address(this));
+
+      if (assets[i] == WETH) {
+        balance += address(this).balance;
+      }
+
+      // Approximate claimable by multiplying
+      // current asset balance with current asset price in USD
+      claimable += balance
+        .mul(uint256(oracle.latestAnswer()))
+        .mul(_allocPCT)
+        .div(10**8)
+        .div(TOTAL_PCT);
+    }
+  }
+
+  /**
+   * @notice
+   * Swaps _amountIn of _asset into distributionToken
+   * @param _asset asset to swap from
+   * @param _amountIn amount to swap of asset
+   * @param _minAmountOut min amount out for every asset type swap
+   */
+  function _swap(
+    address _asset,
+    uint256 _amountIn,
+    uint256 _minAmountOut
+  ) internal {
+    TransferHelper.safeApprove(_asset, address(UNIV3_SWAP_ROUTER), 0);
+    TransferHelper.safeApprove(_asset, address(UNIV3_SWAP_ROUTER), _amountIn);
+
+    ISwapRouter.ExactInputParams memory params = ISwapRouter.ExactInputParams({
+      path: intermediaryPath[_asset],
+      recipient: address(this),
+      amountIn: _amountIn,
+      amountOutMinimum: _minAmountOut
+    });
+
+    // Executes the swap.
+    UNIV3_SWAP_ROUTER.exactInput(params);
+  }
+
+  /**
+     * @notice
+     * add asset
+     * @dev Can be called by admin
+     * @param _asset new asset
+     * @param _oracle ASSET/USD ORACLE.
+     * @param _intermediaryPath path for univ3 swap.
+     * @param _poolFees fees for asset / distributionToken.
+
+     * If intermediary path then pool fee between both pairs
+     * (ex: AAVE / ETH , ETH / USDC)
+     * NOTE: if intermediaryPath empty then single hop swap
+     * NOTE: MUST BE ASSET / USD ORACLE
+     * NOTE: 3000 = 0.3% fee for pool fees
+     */
+  function setAsset(
+    address _asset,
+    address _oracle,
+    address[] calldata _intermediaryPath,
+    uint24[] calldata _poolFees
+  ) external onlyOwner {
+    require(_asset != address(0), "!_asset");
+    uint256 _pathLen = _intermediaryPath.length;
+    uint256 _swapFeeLen = _poolFees.length;
+
+    // We must be setting new valid oracle, or want to keep as is if one exists
+    require(IChainlink(_oracle).decimals() == 8, "!ASSET/USD");
+    require(_pathLen < 2, "invalid intermediary path");
+    require(_swapFeeLen == _pathLen + 1, "invalid pool fees array length");
+
+    // If not set asset
+    if (oracles[_asset] == address(0)) {
+      assets[lastAssetIdx] = _asset;
+      ++lastAssetIdx;
+    }
+
+    // Set oracle for asset
+    oracles[_asset] = _oracle;
+
+    // Multiple pool swaps are encoded through bytes called a `path`.
+    // A path is a sequence of token addresses and poolFees that define the pools used in the swaps.
+    // The format for pool encoding is (tokenIn, fee, tokenOut/tokenIn, fee, tokenOut)
+    // where tokenIn/tokenOut parameter is the shared token across the pools.
+    if (_pathLen > 0) {
+      intermediaryPath[_asset] = abi.encodePacked(
```

Comment:
We should validate that the path is correct after encoding. See how we do it: https://github.com/ribbon-finance/ribbon-v2/blob/9f6e82f2c8cffcf329662bdaf1986ca0360c3855/contracts/libraries/UniswapRouter.sol#L26

Reply:
there is enough external control, and its done rare enough where I think its fine to not have it to save on gas. We can check beforehand, and even if we are wrong and distribute() reverts we can change path later on.